internal/pkg/ocisif/overlay.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2024, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package ocisif
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	ocitsif "github.com/sylabs/oci-tools/pkg/sif"
+	"github.com/sylabs/sif/v2/pkg/sif"
+	"github.com/sylabs/singularity/v4/pkg/image"
+)
+
+var Ext3LayerMediaType types.MediaType = "application/vnd.sylabs.image.layer.v1.ext3"
+
+// HasOverlay returns whether the OCI-SIF at imgPath has an ext3 writable final
+// layer - an 'overlay'.
+func HasOverlay(imagePath string) (bool, error) {
+	fi, err := sif.LoadContainerFromPath(imagePath,
+		sif.OptLoadWithFlag(os.O_RDONLY),
+		sif.OptLoadWithCloseOnUnload(false),
+	)
+	if err != nil {
+		return false, err
+	}
+	defer fi.UnloadContainer()
+
+	ii, err := ocitsif.ImageIndexFromFileImage(fi)
+	if err != nil {
+		return false, fmt.Errorf("while obtaining image index: %w", err)
+	}
+	ix, err := ii.IndexManifest()
+	if err != nil {
+		return false, fmt.Errorf("while obtaining index manifest: %w", err)
+	}
+
+	// One image only.
+	if len(ix.Manifests) != 1 {
+		return false, fmt.Errorf("only single image data containers are supported, found %d images", len(ix.Manifests))
+	}
+	imageDigest := ix.Manifests[0].Digest
+	img, err := ii.Image(imageDigest)
+	if err != nil {
+		return false, fmt.Errorf("while initializing image: %w", err)
+	}
+
+	layers, err := img.Layers()
+	if err != nil {
+		return false, fmt.Errorf("while getting image layers: %w", err)
+	}
+	mt, err := layers[len(layers)-1].MediaType()
```

Comment:
Should probably sanity check that `len(layers) > 0` before doing this just to be safe?